utils/aplay/resampler.c

```diff
@@ -0,0 +1,499 @@
+	return resampler->src_data.src_ratio;
+}
+
+/**
+ * Change the rate ratio applied by the resampler to adjust for the given new
+ * delay value, always trying to move the delay back towards the target value
+ *
+ * @return Returns true if the rate ratio was changed. */
+bool resampler_update_rate_ratio(
+		struct resampler *resampler,
+		snd_pcm_uframes_t frames_read,
+		snd_pcm_uframes_t delay) {
+
+	bool ret = false;
+
+	/* Update the rate ratio only if at least one period
+	 * has passed since the last update. */
+	if (frames_read > 0) {
+		resampler->input_frames += frames_read;
+		/* Prevent the possibility of integer overflow. */
+		resampler->input_frames %= INTMAX_MAX;
+		if (resampler->input_frames - resampler->last_input_frames < resampler->period)
+			return false;
+		resampler->last_input_frames = resampler->input_frames;
+	}
+
+	if (resampler->target_delay == 0 && !is_timespec_zero(&resampler->reset_ts)) {
+		/* Do not restart adaptive resampling until the delay has had time to
+		 * settle to a new value. */
+		struct timespec ts_wait;
+		struct timespec ts_now;
+		gettimestamp(&ts_now);
+		timespecadd(&resampler->reset_ts, &ts_stabilize, &ts_wait);
+		if (difftimespec(&ts_now, &ts_wait, &ts_wait) < 0) {
+			/* Do not allow the target to be outside the configured range. If
+			 * the actual delay is outside that range then try to move it back
+			 * as quickly as possible. */
+			if (delay > resampler->max_target) {
+				resampler->target_delay = resampler->max_target;
+				resampler->src_data.src_ratio =
+					resampler->nominal_rate_ratio - RESAMPLER_STEP_SIZE * RESAMPLER_MAX_STEPS;
+				resampler->rate_ratio_step_count = RESAMPLER_MAX_STEPS;
+			}
+			else if (delay < resampler->min_target) {
+				resampler->target_delay = resampler->min_target;
+				resampler->src_data.src_ratio =
+					resampler->nominal_rate_ratio + RESAMPLER_STEP_SIZE * RESAMPLER_MAX_STEPS;
+				resampler->rate_ratio_step_count = RESAMPLER_MAX_STEPS;
+			}
+			else {
+				/* Once the delay has returned within the permitted target
+				 * range, zero the reset timestamp to indicate that normal
+				 * adaptive resampling has re-started */
+				resampler->reset_ts.tv_sec = 0;
+				resampler->reset_ts.tv_nsec = 0;
+				resampler->target_delay = delay;
+			}
+			resampler->delay_diff = delay - resampler->target_delay;
+			debug("Adaptive resampling enabled: target delay = %.1fms",
+					1000.0 * resampler->target_delay / resampler->in_rate);
+			return true;
+		}
+		return false;
+	}
+
+	snd_pcm_sframes_t delay_diff = delay - resampler->target_delay;
+	snd_pcm_uframes_t delay_diff_abs = labs(delay_diff);
+
+	/* Reset the resampler whenever the delay exceeds the limit. */
+	if (delay_diff_abs > resampler->max_delay_diff) {
+		/* Reset the resampler if not already done. */
+		if (is_timespec_zero(&resampler->reset_ts)) {
+#if DEBUG
+			if (resampler->target_delay != 0)
+				debug("Resetting resampler: Delay difference limit exceeded: %zu > %zu",
```

Comment:
Why only the debug is under the `if`, but the actual `resampler_reset()` is not? Is that a typo or intended logic?

Reply:
The intention is to silence the debug message that would otherwise appear when the ALSA PCM first starts (ie when its start threshold is reached); we still need to reset the resampler there (to set the initial target delay), but I think that the "Delay difference limit exceeded" message is misleading in that special case. Perhaps an explanatory comment in the code is needed to emphasise that this code path applies both to the (normal) adaptive resampling initial start and the (exception) delay limit exceeded condition.